### DISCUSSION TITLE
Revert Redis Tagged cache implementation

### DISCUSSION THREAD

**Discussion by AidasK** [laravel/framework#49159] [Ideas]
### Laravel Version

10.x & 11.x & 12.x

### PHP Version

8.x

### Description

It's well known that laravel tagged cache was removed from the docs a couple of months ago and it looks like a too harsh of a decision for a feature which was in the core for 10 years. So I would like to propose an alternative approach which would not require a lot of work from the devs.

I suggest to revert laravel redis tagged cache to 9.x version. It had no bugs with consistency. It always worked while flushing the tagged cache and it always returned the fresh data. Of course it had leaks, but it was not causing the application to misbehave. This feature was in laravel core for 10 years now and for most of us it was usable. So old implementation can still be used, but some caution messages should be added to the docs.

So my suggestion is to:
 - Bring back the docs. 
 - Revert implementation for redis to 9.x or even better to use [Symfony Cache](https://symfony.com/doc/current/components/cache.html) which supports tags.
 - In case of 9.x, mention that tagged cache does not flush the tag keys and values and they might build up over time. So tagged cache should not be used with dynamically created tags. E.g. used id should not be used as a tag. Though tag cache won't be a problem if it's used with static code enums or constants as they don't scale infinitely. 
 - In case of 9.x, add a requirement to always use an expiry date with the tagged cache to prevent running out of memory. So instead of using cache forever, you should cache only for one week or a month. 
 
So old tagged cache is still usable, though developers have to be aware of the shortcomings. At current state redis tagged cache is unusable as it does not guarantee consistency so memory leaks are not even a concern if it's not working.

### Steps To Reproduce

https://github.com/laravel/framework/pull/48078

https://github.com/laravel/docs/commit/63dde39470588ab1ad39cea9a592f0838ca5ad47

**Comment by jspicee** [comment]
Revert implementation for redis to 9.x with warnings  would be the best for me

**Comment by binaryfire** [comment]
This was my reply to Taylor in the docs commit:

> We have loads of apps that rely on cache tags.... It's a crucial feature for managing caching in any large scale app, or multi-tenant setups etc.
> 
> If you're open to breaking changes in v11, I think refactoring tags to work how they do in every other platform (i.e. much more simply) would be great. The complexity in Laravel comes from how multiple tags work. The readme of this package explains it well: https://github.com/swayok/alternative-laravel-cache.
> 
> I never understood the logic behind Laravel's implementation. Tags should be independent - i.e. you can tag a single cache object with multiple tags, then purge that object using any of those tags. So you could tag a post with posts and tenant_1, and if you purged either the `posts` _or_ `tenant_1` tag that object would be removed.

**Reply by AidasK** [reply]
Wow, was not aware this goodness existed. Though php-cache project seems to be abandoned and they offer to use Symfony Cache instead, which does support tags! It looks like we hit the jackpot @taylorotwell, you won't need to support it and it already works as a drop in replacement 🥳 🎈 
> Back in 2016, this was the first library supporting PSR-6. We created Symfony bundles and made many great libraries in the PHP-cache organization. This was a few years ago and the library is not activly maintained. Starting a new project one should consider using the more performant and activly supported [Symfony Cache](https://symfony.com/doc/current/components/cache.html).

**Reply by binaryfire** [reply]
Hey @AidasK. I'd avoid using that package. Have a look at some of the Laravel-related issues: https://github.com/swayok/alternative-laravel-cache/issues. It doesn't get many updates either. I just linked it because the readme does a good job of explaining the issues with the tagging in core.

IMHO using a third party package for something as fundamental as caching is a very bad idea. You'd be building your app on shaky foundations.

**Reply by AidasK** [reply]
@binaryfire You have missed the point, swayok/alternative-laravel-cache is deprecated and we should use https://symfony.com/doc/current/components/cache.html instead. Laravel already use many of symfony packages anyways so it's safe to use

**Reply by eldair** [reply]
@AidasK do you perhaps have an example of integrating symfony cache into laravel or do you keep and call it separately from laravels' `cache` calls?

**Reply by AidasK** [reply]
Laravel follows https://www.php-fig.org/psr/psr-16/ e.g. SimpleCache interface while symfony follows PSR-6, which is compatible with psr-16. So it should not be hard to write a driver as explained here https://laravel.com/docs/11.x/cache#writing-the-driver

As for tags, laravel has not documented how to extend them, but it's actually quite simple. Instead of extending the `Illuminate\Contracts\Cache\Store;` contract, you should extend Illuminate\Cache\TaggableStore and override tags method with symfony tags implementation. 

So it's actually not that hard to solve and IMO this solution should be a good replacement for laravel native cache implementation.

**Reply by eldair** [reply]
This is what I came up with today, it might be helpful to someone (ofc every input is appreciated) https://github.com/eldair/laravel_symfony_cache/tree/main/extensions/SymfonyRedisCache
I took some ideas from https://github.com/1FF/laravel-mongodb-cache

**Reply by eldair** [reply]
I had to make changes to clean keys because laravel doesn't follow ether psr-6 or psr-16 when it comes to cache keys - laravel uses `:` for various internal cache keys (like maintenance mode) which is not allowed by both PSRs.

**Reply by gerardnll** [reply]
Thanks for the implementation! Is it me or setting \INF when calling ->get() (https://github.com/eldair/laravel_symfony_cache/blob/main/extensions/SymfonyRedisCache/RedisTaggedCache.php#L64) makes the cache revalidate always? Maybe you want to set 0 although that would disable the probabilistic early expiration.

**Reply by eldair** [reply]
It should recalculate since it is in `put` method, if it doesn't recalculate and the item with that key already exists it will not be updated to the desired value from put method.

**Comment by skills-up** [comment]
Has a decision been reached on this yet?

I'm a developing a multi-tenant SaaS application, and imagine my surprise when the tags suddenly went missing from the docs.
At this point, I'm unsure of their availability in the framework, and afraid to use them even if supported.

A tagged cache is required for some significant chunks of data, that are rarely updated. I would like to cache them all forever, and flush the cache only when any corresponding entry is updated.

**Reply by henzeb** [reply]
Not sure how you've set them up. if you have multiple tags, I can't help you atm, but maybe you could try https://packagist.org/packages/henzeb/laravel-cache-index

**Reply by wilsenhc** [reply]
Thank you @henzeb, just saved my life! I will take a better look at the source code later and contribute with anything I can!

**Comment by byErikas** [comment]
bumping this

**Comment by pelmered** [comment]
I really hope this is being worked on for the Laravel 12 release. It is a critical feature for a lot of applications. 
Something based on `symfony/cache` with Laravelified API would be a great solution IMO.

**Comment by therecluse26** [comment]
Why are there still crickets from the Laravel team on this? At the very least, tell us ye or nay if you're planning updates to this functionality, or if we should pound sand and find an alternate solution

**Comment by byErikas** [comment]
since this seems a little dead in the water, I made a solution that implements cache in the `classic` way, but it has possible pitfalls. You're free to try it but read the Readme beforehand. https://github.com/byErikas/cache-tags

It works for my needs, doesn't override too much logic too.

**Reply by onlime** [reply]
> since this seems a little dead in the water, I made a solution that implements cache in the `classic` way, but it has possible pitfalls. You're free to try it but read the Readme beforehand. https://github.com/byErikas/cache-tags
> 
> It works for my needs, doesn't override too much logic too.

Wow, super pumped about this solution! 👏❤️  Hope I can soon give it a try. Would fix so many problems with the way Redis cache tags were implemented since L10.

Thanks so much for your efforts! I'm in the midst of launching a large Laravel project after working 4+ yrs on it, so it's probably not the right time to dig too deep into this. But I hope I'll soon find time for it in the new year. Until then, I hope your package catches a lot of interest from others. Redis/Valkey deserves to be treated better again in Laravel!

**Reply by skills-up** [reply]
I think this is exactly what was required. Although is there a way to create separate namespaces or other methods of scoping here? I would like to scope cache keys by tenant for a multi-tenant app.

**Reply by byErikas** [reply]
> I think this is exactly what was required. Although is there a way to create separate namespaces or other methods of scoping here? I would like to scope cache keys by tenant for a multi-tenant app.

There is. The way I attempted things when developing this was creating a key that's depicted using this behavior: 

`Cache::tags(["tag1", "tag2"])->put("scopedKey", true);` would result in a cache key like this: `tags:tag1.tag2:key:scopedKey` with the value of true.
The key would get put into two key sets, a key set for `tags:tag1` and for `tags:tag2`. This did work, you could flush by a single tag, and retrieve the scoped key by a single tag. However when retrieving items, it was always necessary to perform a scan on the Redis DB to get all the items in the key set, matching the keys to the provided key, and only then if there is a match retrieving and returning it. The same applies to all operations. Overwriting a key, incrementing, and so on. This resulted in a HUGE performance decrease (something like 10x and more at times) and wasn't feasible. 

I wasn't able to come up with a better solution, so here we are, but am always open to ideas and suggestions. 😄

**Reply by binaryfire** [reply]
Just bear in mind that Redis doesn't automatically remove empty sets, and there's no built-in expiry for set members. So you're going to run into memory leak issues with that approach: https://dev.to/zanichelli/investigating-redis-memory-leaks-due-to-laravel-cache-tags-4hd1.

Laravel's Redis tagged cache implementation stores expiry times in the sets as scores + uses a scheduled cleanup command to remove old set members / empty sets. You'll need to implement the same thing to avoid memory issues. That's what we did with our initial internal solution. 

However we recently stopped using Redis sets altogether. We now use a simple `tenant_id:cache:tag:key` prefix system combined with a helper that wraps `Cache` and has fluent methods to scope cache by the tenant + a single tag. Purging is done using scan with an iterator. We realised we never actually retrieving multiple items by tag. Retrieval was always by key and tags were only used for invalidation. For anyone else only using tags for purging and can get away with fixed scopes (eg. tenant id) + a single tag, I'd highly recommend skipping sets and just using prefixes. That way you avoid all the workarounds that are needed for set cleanup.

**Reply by binaryfire** [reply]
FYI here is the issue regarding the memory leak and the PR that solved it:

https://github.com/laravel/framework/issues/43850
https://github.com/laravel/framework/pull/45690

Taylor has said tags won't be removed from core - they're just no longer documented. Which means they'll make sure the existing implementation keeps working but won't be modifying or extending it. So I'm not sure if it's worth re-implementing and maintaining all that logic in a custom implementation. Just my 2c.

**Reply by gerardnll** [reply]
How do you purge keys by tags with a prefix? If you tagged with 2 tags, and use 1 to purge?

**Reply by henzeb** [reply]
Tags are essentially prefixes. You need all tags in the same order you used them when adding them...

**Reply by binaryfire** [reply]
I updated my post to clarify:

> For anyone else only using tags for purging and can get away with fixed scopes (eg. tenant id) + a single tag, I'd highly recommend skipping sets and just using prefixes

As I said, it won't suit everyone. But if you're using tags for scopes that are applied to every entry (eg. a tenant id) then that's always part of the key. In our case, it's the first segment. So we have a fluent method specifically for that. Eg. 
`$cacheHelper->forTenant($tenantId)->tag('posts')->flush()`.

As for how to do it - you'll need to implement that yourself using Redis's [`scan` command](https://github.com/phpredis/phpredis?tab=readme-ov-file#scan). Be sure to use an iterator to avoid long locks. 

If fixed scopes + one tag aren't enough (or you need to retrieve multiple items by tag) then you need to use sets in combination with scheduled cleanups. Definitely doable but a lot more complicated. Redis sets were never intended to be used for ephemeral tagging. If they were, they would have included a member expiry feature.

**Reply by skills-up** [reply]
My usecase is very simple, and I guess very common. I've a multi tenant system, where each tenant has their own cached items, with common entity (and thus key) names. I need to scope key by tenant to avoid cross read/write. The primary use of cache here is to cache large reports that require a lot of calculations. These reports caches are tagged by all contributing entity names, so that when an entity changes I'll purge cache tagged with that entity name, and expect all corresponding caches to be cleared. Almost everything is cached forever as well.

**Reply by binaryfire** [reply]
> These reports caches are tagged by all contributing entity names, so that when an entity changes I'll purge cache tagged with that entity name, and expect all corresponding caches to be cleared.

Laravel's cache tagging never worked like that (although I agree it should have - see my comment [here](https://github.com/laravel/framework/discussions/49159#discussioncomment-8258454)). If you tag an item with 3 tags, you can't purge that item using just one of the tags. You need to pass in all 3.

If you want "tag-with-many / purge-by-any" you're going to need something like Erikas' solution. But it needs set garbage collection. It shouldn't be too difficult to port over the logic from https://github.com/laravel/framework/pull/45690.

**Reply by byErikas** [reply]
> > These reports caches are tagged by all contributing entity names, so that when an entity changes I'll purge cache tagged with that entity name, and expect all corresponding caches to be cleared.
> 
> Laravel's cache tagging never worked like that (although I agree it should have - see my comment [here](https://github.com/laravel/framework/discussions/49159#discussioncomment-8258454)). If you tag an item with 3 tags, you can't purge that item using just one of the tags. You need to pass in all 3.
> 
> If you want "tag-with-many / purge-by-any" you're going to need something like Erikas' solution. But it needs set garbage collection. It shouldn't be too difficult to port over the logic from #45690.

Laravel has garbage cleaning mostly taken care of. From my understanding, `artisan cache:prune-stale-tags` gets tag sets, checks if inner items are expired, and, if so, wipes the set. I have the job set to run every once a day, and memory use seems reasonable.

**Reply by binaryfire** [reply]
If you've made your implementation compatible with [flushStaleTags()](https://github.com/laravel/framework/blob/b9df1f7294bbb7a88311090c8a4782fc92bf46b1/src/Illuminate/Cache/RedisStore.php#L258) then yeah, the default prune command should work. Might be worth adding some test cases to your package to make sure? That's something you really want to ensure is working properly.

**Reply by skills-up** [reply]
> > These reports caches are tagged by all contributing entity names, so that when an entity changes I'll purge cache tagged with that entity name, and expect all corresponding caches to be cleared.
> 
> Laravel's cache tagging never worked like that (although I agree it should have - see my comment [here](https://github.com/laravel/framework/discussions/49159#discussioncomment-8258454)). If you tag an item with 3 tags, you can't purge that item using just one of the tags. You need to pass in all 3.
> 
> If you want "tag-with-many / purge-by-any" you're going to need something like Erikas' solution. But it needs set garbage collection. It shouldn't be too difficult to port over the logic from #45690.

I think this explain certain bugs I had struggled to fix a few years back 😢

**Comment by taylorotwell** [comment]
I would be open to solving the Redis "forever" leaking problem if someone has a clean way to do it. The only way I know of to solve it is by maintaining a list of entries somewhere in the cache?

**Reply by pelmered** [reply]
That sounds great! 

I think something on top of [Symfony Cache](https://symfony.com/doc/current/components/cache.html) that has [good support for tags](https://symfony.com/doc/current/components/cache/cache_invalidation.html) would be a good way forward.
I guess they have solved the memory leak problem as well?

**Comment by doxroot** [comment]
for people who need something very simple for Redis 
```
https://github.com/doxroot/laravel-redis-cache
```
is just a simple wrapper, Laravel already come with all needed for this. However, i use this code above to have a simple way of deleting multiple redis keys if i need.
is not a real tag implementation, again is just a wrapper around powerful Laravel cache system. 

Some apps may need something like tag-name:filter-name:key
So with Redis then is simple to delete all keys like tag-name:* or tag-name:filter-name:*

Not something fancy, but it works for my needs. (hope it can help others)

**Comment by marius-ciclistu** [comment]
Why choose cache when search engines exist and can give you fresh data on each request? Dashboards work well with ElasticSearch/OpenSearch for example. You clone your data in ES from DB in background and when you need it, you aggregate it from ES. If you really want to not hit ES each second if a user refreshes the page, you just wrap the Elasticsearch query in a simple, untagged cache with a tiny TTL of 5 to 10 seconds.

**Comment by macropay-solutions** [comment]
We gave it a try to fix the tagged cache https://github.com/macropay-solutions/maravel-framework/pull/103

**Comment by charescape** [comment]
Considering rapid implementation along with long-term stability and maintainability, it might be a better solution to build Laravel's cache based on the `composer require symfony/cache` component. It is recommended to introduce `symfony/cache` in Laravel 14.
